# Lab 1 - SQL

*Objective:* you will learn to write SQL queries and practice using
[`SQLite`]() and [`Jupyter`]()

## Background

A database for registration of courses, students, and course results
has been developed. The purpose was not to develop a new Ladok
database (Swedish university student database), so everything has been
simplified as much as possible. A course has a course code (e.g.,
EDA216), a name (e.g., Database Technology), a level (G1, G2 or A) and
a number of credits (e.g., 7.5). Students have a social security
number (personnummer) and a name. When a student has passed a course
his/her grade (3, 4 or 5) is registered in the database.

We started by developing an E/R model of the system (E/R stands for
Entity-Relationship). This model is developed in the same way as when
you develop the static model in object-oriented modeling, and you draw
the same kind of UML diagrams. You may instead use traditional E/R
notation, as in the text book. However, diagrams in the traditional
notation take more paper space, and the notation is not fully
standardized, so we will only use the UML notation. The model looks
like this in the different notations (the UML diagram is at the top):

![There should be a image here](lab1.png)

The E/R model is then converted into a database schema in the
relational model. We will later show how the conversion is performed;
for now we just show the final results. The entity sets and the
relationship have been converted into the following relations (the
primary key of each relation is italicized):

 + Students(*ssn*, first_name, last_name)
 + Courses(*course_code*, course_name, level, credits)
 + TakenCourses(*ssn, course_code*, grade)

Examples of instances of the relations:

~~~ {.text}
ssn           first_name   last_name
861103–2438   Bo           Ek
911212–1746   Eva          Alm
950829–1848   Anna         Nyström
...           ...          ...

course_code   course_name                   level    credits
EDA016        Programmeringsteknik          G1       7.5
EDAA01        Programmeringsteknik - FK     G1       7.5
EDA230        Optimerande kompilatorer      A        7.5
...           ...                           ...      ...

ssn           course_code   grade
861103–2438   EDA016        4
861103–2438   EDAA01        3
911212–1746   EDA016        3
...           ...           ...
~~~


The tables have been created with the following SQL statements:

~~~ {.sql}
CREATE TABLE students (
    ssn          char(11),
    first_name   varchar(20) not null,
    last_name    varchar(20) not null,
    primary key  (ssn)
);

CREATE TABLE courses (
    course_code   char(6),
    course_name   varchar(70) not null,
    level         char(2),
    credits       double not null check (credits > 0),
    primary key   (course_code)
);

CREATE TABLE taken_courses (
    ssn           char(11),
    course_code   char(6),
    grade         integer not null check (grade >= 3 and grade <= 5),
    primary key   (ssn, course_code),
    foreign key   (ssn) references students(ssn),
    foreign key   (course_code) references courses(course_code)
);
~~~


All courses that were offered at the Computer Science and Engineering
program at LTH during the academic year 2013/14 are in the table
Courses. Also, the database has been filled with invented data about
students and what courses they have taken. SQL statements like the
following have been used to insert the data:

~~~ {.sql}
INSERT INTO students      VALUES (’861103-2438’, ’Bo’, ’Ek’);
INSERT INTO courses       VALUES (’EDA016’, ’Programmeringsteknik’, ’G1’, 7.5);
INSERT INTO taken_courses VALUES (’861103-2438’, ’EDA016’, 4);
~~~


## Assignments

In [1]:
%load_ext sql

In [2]:
%sql sqlite:///lab1.db

'Connected: None@lab1.db'

The tables `students`, `courses` and `taken_courses` already exist in
your database. If you change the contents of the tables, you can
always recreate the tables with the following command (at the mysql
prompt):

~~~ {.sh}
sqlite3 lab1.db < setup-lab1-db.sql
~~~


After most of the questions there is a number in brackets. This is the
number of rows generated by the question. For instance, [72] after
question a) means that there are 72 students in the database.

a) What are the names (first name, last name) of all the students?
   [72]

In [3]:
%%sql
SELECT * 
FROM students

Done.


[('950705-2308', 'Anna', 'Johansson'),
 ('930702-3582', 'Anna', 'Johansson'),
 ('911212-1746', 'Eva', 'Alm'),
 ('910707-3787', 'Eva', 'Nilsson'),
 ('931213-2824', 'Elaine', 'Robertson'),
 ('951122-1048', 'Maria', 'Nordman'),
 ('910308-1826', 'Helena', 'Troberg'),
 ('941003-1225', 'Lotta', 'Emanuelsson'),
 ('950829-1848', 'Anna', 'Nyström'),
 ('860819-2864', 'Maria', 'Andersson'),
 ('870915-2742', 'Marie', 'Persson'),
 ('900227-2369', 'Märit', 'Aspegren'),
 ('931208-3605', 'Ylva', 'Jacobsson'),
 ('951004-2346', 'Frida', 'Weidel'),
 ('870909-3367', 'Caroline', 'Olsson'),
 ('910915-2068', 'Karolin', 'Ek'),
 ('860206-1065', 'Eva', 'Hjort'),
 ('900528-1540', 'Sofia', 'Kjellberg'),
 ('930502-2040', 'Tina', 'Lööf'),
 ('950317-3520', 'Lisa', 'Berg'),
 ('891021-1287', 'Lotta', 'Bergman'),
 ('880620-2564', 'Susanne', 'Dahl'),
 ('940802-3861', 'Birgit', 'Ewesson'),
 ('850706-2762', 'Ulrika', 'Jonsson'),
 ('890622-1928', 'Jenny', 'Lundin'),
 ('910226-3409', 'Karin', 'Östberg'),
 ('861103-2438', 'Bo', 'Ek'),
 ('931225-3158', 'Bo', 'Ek'),
 ('850819-2139', 'Bo', 'Ek'),
 ('891220-1393', 'Erik', 'Andersson'),
 ('900313-2257', 'Erik', 'Andersson'),
 ('940825-1934', 'Filip', 'Gustavsson'),
 ('860323-1071', 'Niklas', 'Andersson'),
 ('920308-3854', 'David', 'Carlsson'),
 ('890820-2496', 'Tobias', 'Enberg'),
 ('891007-3091', 'Johan', 'Lind'),
 ('920623-3258', 'Magnus', 'Pettersson'),
 ('850517-2597', 'Filip', 'Persson'),
 ('890621-3057', 'Ola', 'Nilsson'),
 ('920812-1857', 'Jakob', 'Malmberg'),
 ('850208-1213', 'Henrik', 'Berg'),
 ('930420-2833', 'John', 'Lind'),
 ('930716-3933', 'Patrik', 'Lundh'),
 ('940801-2971', 'Mikael', 'Nilsson'),
 ('931128-1333', 'Joakim', 'Nilsson'),
 ('940318-2373', 'Conny', 'Modig'),
 ('881030-2772', 'Christian', 'Wallman'),
 ('931015-1271', 'Magnus', 'Strömgren'),
 ('900206-3735', 'Daniel', 'Ahlman'),
 ('891231-2554', 'Lars', 'Molin'),
 ('900526-1450', 'Magnus', 'Nilsson'),
 ('950830-2637', 'Anders', 'Olsson'),
 ('880322-3951', 'Martin', 'Alm'),
 ('950714-2190', 'Håkan', 'Mossgren'),
 ('911020-3538', 'Fredrik', 'Ryd'),
 ('921222-2113', 'Mats', 'Wikström'),
 ('891106-1277', 'Anders', 'Magnusson'),
 ('920906-3371', 'Andreas', 'Molin'),
 ('890103-1256', 'Johan', 'Brattberg'),
 ('920514-3433', 'Johan', 'Olsson'),
 ('911015-3758', 'Jonathan', 'Jönsson'),
 ('950125-1153', 'Magnus', 'Hultgren'),
 ('950715-2090', 'Axel', 'Nord'),
 ('900918-3999', 'Henrik', 'Hult'),
 ('880206-1915', 'Joakim', 'Hall'),
 ('900129-3374', 'Henrik', 'Gustavsson'),
 ('881110-1272', 'Torbjörn', 'Modig'),
 ('921029-1995', 'Daniel', 'Axelsson'),
 ('901030-1895', 'Per-Erik', 'Pettersson'),
 ('920921-2499', 'Roger', 'Brorsson'),
 ('930804-3050', 'Patrik', 'Ohlsson'),
 ('910101-1234', 'Peter', 'Solberg')]

b) Same as question a) but produce a sorted listing. Sort first by
   last name and then by first name.

In [ ]:
%%sql


c) Which students were born in 1985? [4]

In [ ]:
%%sql


d) What are the names of the female students, and which are their
   social security numbers? The next-to-last digit in the social
   security number is even for females. The SQLite function
   `substr(str,m,n)` returns `n` characters from the string `str`,
   starting at character `m`. [26]

In [ ]:
%%sql


e) How many students are registered in the database?

In [ ]:
%%sql


f) Which courses are offered by the department of Mathematics (course
   codes `FMAxxx`)? [22]

In [ ]:
%%sql


g) Which courses give more than 7.5 credits? [16]

In [ ]:
%%sql


h) How may courses are there on each level `G1`, `G2` and `A`?

In [ ]:
%%sql


i) Which courses (course codes only) have been taken by the student
   with social security number 910101–1234? [35]

In [ ]:
%%sql


j) What are the names of these courses, and how many credits do they give?

In [ ]:
%%sql


k) How many credits has the student taken?

In [ ]:
%%sql


l) Which is the student’s grade average (arithmetic mean, not weighted) on the courses?

In [ ]:
%%sql


m) Same questions as in questions i)–l), but for the student Eva Alm. [26]

In [ ]:
%%sql


n) Which students have taken 0 credits? [11]

In [ ]:
%%sql


o) Which students have the highest grade average? Advice: define and
   use a view that gives the social security number and grade average
   for each student.

In [ ]:
%%sql


p) List the social security number and total number of credits for all
   students. Students with no credits should be included with 0
   credits, not null. If you do this with an outer join you might want
   to use the function coalesce(v1, v2, ...); it returns the first
   value which is not null. [72]

In [ ]:
%%sql


q) Is there more than one student with the same name? If so, who are
   these students and what are their social security numbers? [7]

In [ ]:
%%sql
